Fix overlay vxlan races

ctelfer opened the pull request:
This function takes the patch offered in https://github.com/docker/libnetwork/pull/1800 and attempts to address the issues found in https://github.com/docker/libnetwork/issues/1765.  It removes a race condition documented in the issue where "re-once"ing the creation of an overlay sandbox can race with incoming join requests to cause a leak/collision with the vxlan interface.  This PR addresses said issue by removing the "re-once" pattern and replacing it with a traditional mutex+boolean initializer pattern.  This also allows removing some restore error handling by having the sandbox join/leave perform proper cleanup on error.  More information is available in the commit logs of the PR.

selansen commented:
CI Failed

fcrisciani commented on drivers/overlay/ov_network.go:
```diff
@@ -296,48 +295,62 @@ func (d *driver) RevokeExternalConnectivity(nid, eid string) error {
-	n.joinCnt++
-}
+func (n *network) joinSandbox(s *subnet, restore bool, incJoinCnt bool) error {
+	networkOnce.Do(networkOnceInit)
```
if 2 joinSanbox are happening, the first one will trigger this `networkOnceInit`, while the second one will continue forward acquiring the lock. Is it ok for the second one to go ahead also before the first `networkOnceInit` is completed?
If I look at the code I guess this will create a potential race with the `n.initSandbox(restore)` that actually if restore is `false` will try to delete the VNI
Am I missing something?

ctelfer replied:
I checked on this very early on...  once.Do() holds a mutex while doing the initialization.  If two goroutines execute the same once.Do() concurrently, the second one will block until the first one completes.

fcrisciani commented on drivers/overlay/ov_network.go:
```diff
@@ -296,48 +295,62 @@ func (d *driver) RevokeExternalConnectivity(nid, eid string) error {
-	})
-	return s.initErr
+	subnetErr := s.initErr
+	if !s.sboxInit {
```
isn't this dead code? 
if at line 304 was false, will be set to true
if it was already true then won't enter here anyway

ctelfer replied:
This is `s.sboxInit` not `n.sboxInit` ...  So no, this is different. :)

fcrisciani commented on drivers/overlay/ov_network.go:
```diff
@@ -580,6 +593,23 @@ func (n *network) setupSubnetSandbox(s *subnet, brName, vxlanName string) error
+		// failure of vxlan device creation if the vni is assigned to some other
+		// network.
+		if deleteErr := deleteInterface(vxlanName); deleteErr != nil {
+			logrus.Warnf("could not delete vxlan interface, %s, error %v, after config error, %v\n", vxlanName, deleteErr, err)
```
you can remove the trailing `\n`

ctelfer replied:
ah, yes, will do.

selansen commented on drivers/overlay/ov_network.go:
```diff
@@ -296,48 +295,62 @@ func (d *driver) RevokeExternalConnectivity(nid, eid string) error {
+			s.sboxInit = true
+		}
+	}
+	if subnetErr != nil {
```
We declare subnetErr inside if !s.sboxInit { . and again using it for outside the if scope.  Can we declare outside of if and make the scope to function level.

ctelfer replied:
Good catch.  That's definitely a bug.  It is not supposed to be redeclared within the 'if'.

ctelfer commented:
Pushed an update with a much simplified locking scheme.   There is now only a single lock for each network and all its subnets.   I originally put in the second lock because of a deadlock (which I had hit during testing before the initial PR) described in the commit log.  This new version breaks the deadlock by posting the notification to initialize the peerDB on a join using a fresh goroutine.  This prevents the 'join' from deadlocking waiting on the channel of the peerDB goroutine while the peerdb goroutine is waiting for the 'join' to release the network lock.

This version also addresses the comments by @fcrisciani (newlines) and @selansen (redeclaration of subnetErr).

fcrisciani commented on drivers/overlay/ov_network.go:
```diff
@@ -296,29 +294,39 @@ func (d *driver) RevokeExternalConnectivity(nid, eid string) error {
 
-func (n *network) incEndpointCount() {
+func (n *network) joinSandbox(s *subnet, restore bool, incJoinCount bool) error {
+	networkOnce.Do(networkOnceInit)
```
can you keep this comment:
```
	// If there is a race between two go routines here only one will win
	// the other will wait.
```

fcrisciani commented on drivers/overlay/ov_network.go:
```diff
@@ -296,29 +294,39 @@ func (d *driver) RevokeExternalConnectivity(nid, eid string) error {
-		s.initErr = n.initSubnetSandbox(s, restore)
-	})
-	return s.initErr
+	subnetErr := s.initErr
```
we can simply avoid using this subnetErr, and just leverage the s.initErr

fcrisciani replied:
also for line 321

ctelfer replied:
Actually, there is a very specific reason why it isn't.  We want to return:
  - `s.initErr` *iff* it was previously set
  - otherwise we want to return the result of `n.initSubnetSandbox(s, restore)` *regardless* of whether that value gets stored in `s.initErr`.  So, on line 321 we do not want to just return `s.initErr`.  It may not have been set if there was a failure but we are not in a `restore` case.

This is the semantics from the https://github.com/docker/libnetwork/pull/1800 and I've preserved them.  The logic basically implies that if an error is recoverable, return it so it is flagged appropriately, but to make it persistent.  But if the error is not recoverable (restore case), make it persistent.

fcrisciani replied:
do we expect that the error failure will be different? on multiple retry?

fcrisciani commented on drivers/overlay/ov_network.go:
```diff
@@ -470,7 +477,7 @@ func (n *network) generateVxlanName(s *subnet) string {
 	}
 
-	return "vx-" + fmt.Sprintf("%06x", n.vxlanID(s)) + "-" + id
+	return "vx-" + fmt.Sprintf("%06x", s.vni) + "-" + id
```
considering that we are touching this line, this will be more efficient:
`fmt.Sprintf("vx-%06x-%v", s.vni,id)`

fcrisciani commented on drivers/overlay/ov_network.go:
```diff
@@ -483,7 +490,7 @@ func (n *network) generateBridgeName(s *subnet) string {
 
 func (n *network) getBridgeNamePrefix(s *subnet) string {
-	return "ov-" + fmt.Sprintf("%06x", n.vxlanID(s))
+	return "ov-" + fmt.Sprintf("%06x", s.vni)
```
same here just with a single Sprintf

selansen commented on drivers/overlay/ov_network.go:
```diff
@@ -1031,22 +1036,25 @@ func (n *network) releaseVxlanID() ([]uint32, error) {
 
-		n.setVxlanID(s, 0)
+	for _, vni := range vnis {
+		n.driver.vxlanIdm.Release(uint64(vni))
```
why do we need  extra for loop instead of using previous logic where its all done inside one place. What are we trying to achieve by doing this ?

ctelfer replied:
Fair question.  The vxlanIdm.Release() operation can do locking and write to store while other overlay networks are coming and going as well.   So, I was thinking that this prevents holding the network lock through all those other potentially blocking operations.  (i.e. collect the work list using the lock and then go through the process of releasing them)   But since this function only gets called while the driver is holding the driver lock, that may be a superfluous optimization.

selansen commented on drivers/overlay/ov_network.go:
```diff
@@ -1031,22 +1036,25 @@ func (n *network) releaseVxlanID() ([]uint32, error) {
 func (n *network) obtainVxlanID(s *subnet) error {
 	//return if the subnet already has a vxlan id assigned
-	if s.vni != 0 {
+	if n.vxlanID(s) != 0 {
```
Other places you replaced n.vxlanID(s) with s.vni or n.vni.   Here we are using n.vxlanID(s) instead of directly accessing with s.vni . Trying to understand why this change is required ?

ctelfer replied:
Also fair.   The revamped locking holds the network lock through network.join() and network.leave() operations.  As such, calling `n.vxlanID(s)` within those methods (or their sub-functions) would be a double lock.  That's the reasons for the changes above.

However, the changes above also *remove* the subnet locks and place subnet locking all within the domain of the network's lock.   As such, `network` methods need to acquire the network lock before accessing any of the subnet fields.  This driver calls obtainVxlanID() outside of the context of `network.join()` in several places.  Hence accessing the subnet members requires doing so through the network lock.

The lack of locking in obtainVxlanId() was, in principle, a race condition before that just wasn't addressed.

selansen commented on drivers/overlay/ov_network.go:
```diff
@@ -1059,7 +1067,7 @@ func (n *network) obtainVxlanID(s *subnet) error {
 		}
 
-		if s.vni == 0 {
+		if n.vxlanID(s) == 0 {
```
same as above

fcrisciani commented on drivers/overlay/ov_network.go:
```diff
@@ -861,33 +878,19 @@ func (d *driver) getNetworkFromStore(nid string) *network {
 
 func (n *network) setVxlanID(s *subnet, vni uint32) {
 	n.Lock()
+	defer n.Unlock()
```
the defer is just slower, in this case for a 3 lines methods, I would keep it as before

selansen commented:
LGTM

ctelfer commented:
Addressed the comment, defer and sprintf comments above and have re-pushed.

ctelfer commented:
Rebased to head of master, updated a few comments and re-tested a few times to make sure this looks good.  Everything has been clean.

I'll admit that I'm still a bit nervous after https://github.com/docker/libnetwork/pull/2143 / https://github.com/docker/libnetwork/pull/2180.  But from a rational standpoint, what this PR does is remove the unsafe re-onceing from the overlay code, make joins and leaves to the overlay sandboxes atomic, and merge in  https://github.com/docker/libnetwork/pull/1800's error handling code.  It should be safe to merge.

m4r10k commented:
@fcrisciani do you know which Docker version will include this merge (vendors this libnetwork hash)? We are running 18.03 and are regularly experiencing the issue that the vxlan file is already existing.
~~~
28686: vx-00101b-nmh1s: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN mode DEFAULT group default 
    vxlan id 4123 srcport 0 0 dstport 4789 proxy l2miss l3miss ageing 300 addrgenmode eui64 
29077: vx-001017-sxegy: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN mode DEFAULT group default 
    vxlan id 4119 srcport 0 0 dstport 4789 proxy l2miss l3miss ageing 300 addrgenmode eui64 
23209: vx-001006-tlzfj: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN mode DEFAULT group default 
    vxlan id 4102 srcport 0 0 dstport 4789 proxy l2miss l3miss ageing 300 addrgenmode eui64 
29632: vx-001011-ox0iz: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN mode DEFAULT group default 
    vxlan id 4113 srcport 0 0 dstport 4789 proxy l2miss l3miss ageing 300 addrgenmode eui64 
29658: vx-001028-jjw67: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN mode DEFAULT group default 
    vxlan id 4136 srcport 0 0 dstport 4789 proxy l2miss l3miss ageing 300 addrgenmode eui64 
28399: vx-00103a-i7t6y: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN mode DEFAULT group default 
    vxlan id 4154 srcport 0 0 dstport 4789 proxy l2miss l3miss ageing 300 addrgenmode eui64 
28911: vx-001020-ups39: <BROADCAST,MULTICAST> mtu 1450 qdisc noop state DOWN mode DEFAULT group default 
    vxlan id 4128 srcport 0 0 dstport 4789 proxy l2miss l3miss ageing 300 addrgenmode eui64 
~~~
